amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/ExchangeMessageRouter.java

```diff
@@ -0,0 +1,428 @@
+        }
+    }
+
+    private int getAvailablePermits() {
+        int availablePermits = replicatorQueueSize - PENDING_SIZE_UPDATER.get(this);
+        if (availablePermits <= 0) {
+            if (log.isDebugEnabled()) {
+                log.debug("{} Replicator queue is full, availablePermits: {}, pause route.",
+                        exchange.getName(), availablePermits);
+            }
+            return 0;
+        }
+        return availablePermits;
+    }
+
+    private void processMessages(List<Entry> entries) {
+        for (Entry entry : entries) {
+            PENDING_SIZE_UPDATER.incrementAndGet(this);
+            Map<String, String> props;
+            MessageImpl<byte[]> message;
+            try {
+                message = MessageImpl.deserialize(entry.getDataBuffer());
+                props = message.getMessageBuilder().getPropertiesList().stream()
+                        .collect(Collectors.toMap(KeyValue::getKey, KeyValue::getValue));
+            } catch (IOException e) {
+                log.error("Deserialize entry dataBuffer failed. exchangeName: {}, skip it first.",
+                        exchange.getName(), e);
+                PENDING_SIZE_UPDATER.decrementAndGet(this);
+                continue;
+            }
+            String routingKey = props.getOrDefault(MessageConvertUtils.PROP_ROUTING_KEY, "").toString();
+            Set<Destination> destinations = getDestinations(routingKey, getMessageHeaders());
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+            byte[] data;
+            if (!destinations.isEmpty()) {
+                data = new byte[message.getDataBuffer().readableBytes()];
+                message.getDataBuffer().readBytes(data);
+                for (Destination des : destinations) {
+                    futures.add(sendMessage(des, data, props));
+                }
+            }
+            final Position position = entry.getPosition();
+            entry.release();
+            FutureUtil.waitForAll(futures).whenComplete((__, t) -> {
+                if (t != null) {
+                    log.error("Failed to route message {}", position, t);
+                    cursor.rewind();
+                    sendComplete();
+                    return;
+                }
+                sendComplete();
+                cursor.asyncDelete(position, new AsyncCallbacks.DeleteCallback() {
+                    @Override
+                    public void deleteComplete(Object ctx) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("{} Deleted message at {}", exchange.getName(), ctx);
+                        }
+                    }
+
+                    @Override
+                    public void deleteFailed(ManagedLedgerException exception, Object ctx) {
+                        log.error("{} Failed to delete message at {}", exchange.getName(), ctx, exception);
+                    }
+                }, entry.getPosition());
```

Comment:
The entry has been released, and the 190 line position variable should be used.

Reply:
nice cache, fixed